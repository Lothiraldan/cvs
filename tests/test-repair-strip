#!/bin/sh

echo "[extensions]" >> $HGRCPATH
echo "mq=">> $HGRCPATH

teststrip() {
    hg -q up -C $1
    echo % before update $1, strip $2
    hg parents
    chmod -$3 $4
    hg strip $2 2>&1 | sed 's/\(saving bundle to \).*/\1/' | sed 's/Permission denied.*\.hg\/store\/\(.*\)/Permission denied \.hg\/store\/\1/'
    echo % after update $1, strip $2
    chmod +$3 $4
    hg verify
    echo % journal contents
    cat .hg/store/journal | sed 's/\.i[^\n]*/\.i/'
    ls .hg/store/journal >/dev/null 2>&1 && hg recover
    ls .hg/strip-backup/* >/dev/null 2>&1 && hg unbundle -q .hg/strip-backup/*
    rm -rf .hg/strip-backup
}

hg init test
cd test

echo a > a
hg -q ci -m "a" -A

echo b > b
hg -q ci -m "b" -A

echo c > c
hg -q ci -m "c" -A

teststrip 0 1 w .hg/store/data/b.i
teststrip 0 1 r .hg/store/data/b.i
teststrip 0 1 w .hg/store/00changelog.i
