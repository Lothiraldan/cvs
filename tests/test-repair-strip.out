% before update 0, strip 2
changeset:   0:cb9a9f314b8b
user:        test
date:        Thu Jan 01 00:00:00 1970 +0000
summary:     a

saving bundle to 
transaction abort!
failed to truncate data/b.i
rollback failed - please run hg recover
abort: Permission denied .hg/store/data/b.i
% after update 0, strip 2
checking changesets
checking manifests
crosschecking files in changesets and manifests
checking files
 b@?: rev 1 points to nonexistent changeset 2
 (expected 1)
 b@?: 736c29771fba not in manifests
2 files, 2 changesets, 3 total revisions
1 warnings encountered!
2 integrity errors encountered!
% journal contents
00changelog.i
00manifest.i
data/b.i
data/c.i
rolling back interrupted transaction
checking changesets
checking manifests
crosschecking files in changesets and manifests
checking files
2 files, 2 changesets, 2 total revisions
% before update 0, strip 2
changeset:   0:cb9a9f314b8b
user:        test
date:        Thu Jan 01 00:00:00 1970 +0000
summary:     a

abort: Permission denied .hg/store/data/b.i
% after update 0, strip 2
checking changesets
checking manifests
crosschecking files in changesets and manifests
checking files
3 files, 4 changesets, 4 total revisions
% journal contents
cat: .hg/store/journal: No such file or directory
% before update 0, strip 2
changeset:   0:cb9a9f314b8b
user:        test
date:        Thu Jan 01 00:00:00 1970 +0000
summary:     a

saving bundle to 
transaction abort!
failed to truncate 00manifest.i
rollback failed - please run hg recover
abort: Permission denied .hg/store/00manifest.i
% after update 0, strip 2
checking changesets
checking manifests
 manifest@?: rev 2 points to nonexistent changeset 2
 manifest@?: rev 3 points to nonexistent changeset 3
crosschecking files in changesets and manifests
 c@?: in manifest but not in changeset
checking files
 b@2: 736c29771fba in manifests not found
 data/c.i@?: missing revlog!
 ?: empty or missing c
 c@3: 149da44f2a4e in manifests not found
3 files, 2 changesets, 2 total revisions
7 integrity errors encountered!
(first damaged changeset appears to be 2)
% journal contents
00changelog.i
00manifest.i
data/b.i
data/c.i
rolling back interrupted transaction
checking changesets
checking manifests
crosschecking files in changesets and manifests
checking files
2 files, 2 changesets, 2 total revisions
